### PR TITLE
Add FNet Encoder Layer

### LINKED ISSUE #40: Add FNet Encoder Layer
@mattdangerw , @chenmoneygithub and the rest of the keras-nlp team:

The [FNet](https://arxiv.org/abs/2105.03824) paper, which came out last year, uses FFTs instead of attention and achieves 95% of BERT's performance on various downstream tasks. It would be great if we could add this to the library. Simultaneously, a text classification example can be added (using a few FNet Encoder layers). Would love to take this up, if the team is interested :).

### PR DESCRIPTION
**Pull request by abheesht17**
Resolves #40. 

I've implemented the FNet Encoder Layer. I'll add unit tests once the code has been finalised. @mattdangerw ,@chenmoneygithub, please go through it. Thanks!

Here is the link to the Colab notebook where I do a forward pass: https://colab.research.google.com/drive/1ldWMSqw8q8qwndHUxtuILxowRRFiBTbT?usp=sharing 

I'm planning to add an example for this (text classification). I'll stack 2-3 encoder layers and train the model.
**Caveat:** In the paper, the authors mention that they used SentencePiece Tokenizer (this hasn't been implemented in `keras-nlp` yet). So, should I add the example later in a separate PR? Or should I just use SentencePiece from TF Text?

### CODE REVIEW

**Comment by abheesht17** [comment]
Hey, @mattdangerw. Any thoughts on this?

File: keras_nlp/layers/fnet_encoder.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+        # Feedforward layer.
+        self._intermediate_dense = keras.layers.Dense(
+            self.intermediate_dim, activation=self.activation
+        )
+        self._output_dense = keras.layers.Dense(feature_size)
+        self._output_dropout = keras.layers.Dropout(rate=self.dropout)
+
+    def _fourier_transform(self, input):
+        # Apply FFT on the input and take the real part. Before we apply fourier
```
Comment (chenmoneygithub) [inline keras_nlp/layers/fnet_encoder.py:100]: Please split this to two comments.

File: keras_nlp/layers/fnet_encoder.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+                of shape [batch_size, sequence_length, feature_dim].
+
+        Returns:
+            A Tensor of the same shape as the `inputs`.
+        """
+
+        if not self._built:
+            self._build(inputs.shape)
+
+        # Apply fourier transform on the input. Note: We don't have padding
```
Comment (chenmoneygithub) [inline keras_nlp/layers/fnet_encoder.py:128]: I want to understand this a bit more - are you saying the original paper does not do padding, and take in RaggedTensor as input? Thanks!

**Review by chenmoneygithub — COMMENTED** [review]
Thanks for the PR! 

Dropped some comments, and please add unit tests to test the layer's funcitonality, thanks!

**Comment by mattdangerw** [comment]
Actually, if it's ok, maybe wait until this Tuesday to fully flesh this out? We have an internal sync meeting then, and will discussing where we want to set our bar (e.g. with citations) for inclusion in this repo.

I'm pretty sure we will want this, but if you wait till then, we could give a more confident green light. Thanks for patience! Really do appreciate the contributions!

**Comment by abheesht17** [comment]
Sure! Thank you :)

File: keras_nlp/layers/fnet_encoder.py (hunk 2 above)
Comment (abheesht17) [reply keras_nlp/layers/fnet_encoder.py:128]: So, as far as I understand, they do pad the sequences so that every sequence is of the same length. See this: https://github.com/google-research/google-research/blob/master/f_net/layers.py#L106. The expected shape is `(bsz, max_seq_length, hidden_dim)`.

However, the padding mask is not used at all. They delete it here: https://github.com/google-research/google-research/blob/master/f_net/layers.py#L114. So, the mixing is done over all the tokens, including the padding tokens. See the input pipeline here: https://github.com/google-research/google-research/blob/master/f_net/input_pipeline.py#L100.

File: keras_nlp/layers/fnet_encoder.py (hunk 1 above)
Comment (abheesht17) [reply keras_nlp/layers/fnet_encoder.py:100]: Done 👍🏼

File: keras_nlp/layers/fnet_encoder.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+    This class follows the architecture of FNet encoder layer in paper
+    "FNet: Mixing Tokens with Fourier Transforms"
+    (https://arxiv.org/abs/2105.03824). Users can instantiate multiple instances
+    of this class to stack up the encoder.
+
+    Args:
+        intermediate_dim: int, defaults to 3072. The hidden size of feedforward
+            network.
+        dropout: float, defaults to 0.1. the dropout value, applied in the
```
[the remark this answers is not in the document]
Comment (abheesht17) [reply keras_nlp/layers/fnet_encoder.py:32]: Done 👍🏼

**Review by abheesht17 — COMMENTED** [review]
Addressed first round of comments.

File: keras_nlp/layers/fnet_encoder.py (hunk 2 above)
Comment (abheesht17) [reply keras_nlp/layers/fnet_encoder.py:129]: I received a reply from the first author, James Lee-Thorp. This is the gist of the message:

**For FNet, padding tokens are applied, but padding tokens are not masked**. There is a possibility that the model may become sensitive to samples with varying number of padding tokens...but in practice, the authors did not notice any major performance issues. So, I think we can keep it the same, we don't need to apply padding masks.

If we do want to apply a padding mask, we should only apply it after the embedding layer, and not after every Fourier sublayer. This is because if we apply it after every Fourier sublayer, we will end up zeroing out some frequencies which is counter-productive.

**Comment by abheesht17** [comment]
Awesome, @mattdangerw! Thank you :). I'll add the the tests and the initialiser arguments. 

One more thing - could you and @chenmoneygithub please give https://github.com/keras-team/keras-nlp/pull/43#discussion_r830591662 and https://github.com/keras-team/keras-nlp/pull/43#discussion_r831721727 a look so that we can have closure on the padding mask issue?

**Comment by abheesht17** [comment]
@mattdangerw, @chenmoneygithub, I've done the following:
- [x] Added UTs (followed the UTs for TransformerEncoder).
- [x] Added kernel and bias initialisers.

Let me know if it's okay.

File: keras_nlp/layers/fnet_encoder.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+        super().__init__(name=name, **kwargs)
+        self.intermediate_dim = intermediate_dim
+        self.dropout = dropout
+        self.activation = keras.activations.get(activation)
+        self.layer_norm_epsilon = layer_norm_epsilon
+        self.kernel_initializer = keras.initializers.get(kernel_initializer)
+        self.bias_initializer = keras.initializers.get(bias_initializer)
+        self._built = False
+
+    def _build(self, input_shape):
```
Comment (fchollet) [inline keras_nlp/layers/fnet_encoder.py:86]: This should just be the built-in `build()`

File: keras_nlp/layers/fnet_encoder.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+            bias_initializer=self.bias_initializer,
+        )
+        self._output_dense = keras.layers.Dense(
+            feature_size,
+            kernel_initializer=self.kernel_initializer,
+            bias_initializer=self.bias_initializer,
+        )
+        self._output_dropout = keras.layers.Dropout(rate=self.dropout)
+
+    def _fourier_transform(self, input):
```
Comment (fchollet) [inline keras_nlp/layers/fnet_encoder.py:113]: In-line it in `call`

File: keras_nlp/layers/fnet_encoder.py [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+        # Apply FFT on the input and take the real part.
+        # Before we apply fourier transform, let's convert the dtype of the
+        # input tensor to complex64.
+        input = tf.cast(input, tf.complex64)
+        mixing_output = tf.math.real(tf.signal.fft2d(input))
+        return mixing_output
+
+    def _add_and_norm(self, input1, input2, norm_layer):
```
Comment (fchollet) [inline keras_nlp/layers/fnet_encoder.py:121]: In-line it in `call`

File: keras_nlp/layers/fnet_encoder.py [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+        return norm_layer(input1 + input2)
+
+    def _feed_forward(self, input):
```
Comment (fchollet) [inline keras_nlp/layers/fnet_encoder.py:124]: In-line it in `call`

File: keras_nlp/layers/fnet_encoder.py [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+    output = model(input_data)
+    ```
+
+    References:
+        [Lee-Thorp et al., 2021](https://arxiv.org/abs/2105.03824)
+    """
+
+    def __init__(
+        self,
+        intermediate_dim=3072,
```
Comment (fchollet) [inline keras_nlp/layers/fnet_encoder.py:68]: Where does this value come from?

File: keras_nlp/layers/fnet_encoder.py [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+        dropout=0.1,
```
Comment (fchollet) [inline keras_nlp/layers/fnet_encoder.py:69]: Should default to 0

File: keras_nlp/layers/fnet_encoder.py [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+        activation="gelu",
+        layer_norm_epsilon=1e-12,
```
Comment (fchollet) [inline keras_nlp/layers/fnet_encoder.py:71]: This seems way too small

File: keras_nlp/layers/fnet_encoder.py [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+        kernel_initializer="glorot_uniform",
```
Comment (fchollet) [inline keras_nlp/layers/fnet_encoder.py:72]: Is this choice of initializer the best practice for this layer?

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

File: keras_nlp/layers/fnet_encoder.py (hunk 8 above)
Comment (mattdangerw) [reply keras_nlp/layers/fnet_encoder.py:68]: Let's remove the default. intermediate_dim should be a required argument to pass in.

File: keras_nlp/layers/fnet_encoder.py (hunk 4 above)
Comment (abheesht17) [reply keras_nlp/layers/fnet_encoder.py:86]: Ah, overriding the `build()` function in `tf.keras.layers.Layer`? Should I make the same changes to the `TransformerEncoder` class? 
https://github.com/keras-team/keras-nlp/blob/master/keras_nlp/layers/transformer_encoder.py#L83

File: keras_nlp/layers/fnet_encoder.py (hunk 5 above)
Comment (abheesht17) [reply keras_nlp/layers/fnet_encoder.py:113]: Done!

File: keras_nlp/layers/fnet_encoder.py (hunk 6 above)
Comment (abheesht17) [reply keras_nlp/layers/fnet_encoder.py:121]: Done!

File: keras_nlp/layers/fnet_encoder.py (hunk 7 above)
Comment (abheesht17) [reply keras_nlp/layers/fnet_encoder.py:124]: Done!

**Review by abheesht17 — COMMENTED** [review]
Thank you for the review comments, @fchollet, @mattdangerw! :)

Summary:

- For all argument-related comments, I've attached links to the config present in the official FNet code. What do you think I should do - keep the same defaults as given in the FNet config, or change it to the defaults given in `TransformerEncoder`?
- Please see the comment on initializers. I've changed everything to `RandomNormal` with a standard deviation of `2e-2`.
- For the comments on making the functions in-line and the one about `build()`, I had followed what's been done in the `TransformerEncoder` class. I've now changed it to what @fchollet mentioned in the comments. Should I make the same changes in `TransformerEncoder` and `TransformerDecoder`?

File: keras_nlp/layers/fnet_encoder.py (hunk 4 above)
Comment (chenmoneygithub) [reply keras_nlp/layers/fnet_encoder.py]: the default `build()` only works if your `call()` method takes one input, the TransformerEncoder takes multiple inputs, so default `build()` cannot be used there.

File: keras_nlp/layers/fnet_encoder.py (hunk 8 above)
Comment (chenmoneygithub) [reply keras_nlp/layers/fnet_encoder.py:68]: At component level let's do not set the default, we will pass this value when we are building the actual model.

File: keras_nlp/layers/fnet_encoder.py [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+
+        def _add_and_norm(input1, input2, norm_layer):
+            return norm_layer(input1 + input2)
+
+        def _feed_forward(input):
+            x = self._intermediate_dense(input)
+            x = self._output_dense(x)
+            return self._output_dropout(x)
+
+        # Apply fourier transform on the input.
```
Comment (chenmoneygithub) [inline keras_nlp/layers/fnet_encoder.py:138]: I would prefer putting these comments into the class' docstring, since the line `mixing_output = _fourier_transform(inputs)` does not indicate this much information.

Additionally, this code reference is error-prone, because the line number is subject to change. Let's delete it, thanks!

File: keras_nlp/layers/fnet_encoder.py (hunk 4 above)
Comment (abheesht17) [reply keras_nlp/layers/fnet_encoder.py]: Ohh, I see. Thanks! :)

File: keras_nlp/layers/fnet_encoder.py (hunk 8 above)
Comment (abheesht17) [reply keras_nlp/layers/fnet_encoder.py]: Ah, alright. I've removed the default value!

File: keras_nlp/layers/fnet_encoder.py (hunk 12 above)
Comment (abheesht17) [reply keras_nlp/layers/fnet_encoder.py:144]: Done!

**Review by abheesht17 — COMMENTED** [review]
Thank you for the comments, @chenmoneygithub! I've addressed them.

File: keras_nlp/layers/fnet_encoder.py (hunk 9 above)
Comment (mattdangerw) [reply keras_nlp/layers/fnet_encoder.py:75]: Right, but we are not shipping a model architecture here, we are shipping a reusable layer. I would be inclined to use the same defaults as the TransformerEncoder block for most of the args here so this is essentially a swap in replacement. Having each encoder block we ship use different defaults everywhere would be a lot to track, and lead to bad usability.

File: keras_nlp/layers/fnet_encoder.py [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+
+        Args:
+            inputs: a Tensor. The input data to TransformerEncoder, should be
+                of shape [batch_size, sequence_length, feature_dim].
+
+        Returns:
+            A Tensor of the same shape as the `inputs`.
+        """
+
+        def _fourier_transform(input):
```
Comment (mattdangerw) [inline keras_nlp/layers/fnet_encoder.py:128]: I don't think there's any reason to keep this inner functions private with underscores. They are local to the call function.

You only need the underscore for self._x to communicate that it's private.

File: keras_nlp/layers/fnet_encoder.py [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+
+        def _feed_forward(input):
+            x = self._intermediate_dense(input)
+            x = self._output_dense(x)
+            return self._output_dropout(x)
+
+        # Apply fourier transform on the input.
+        mixing_output = _fourier_transform(inputs)
+
+        # LayerNorm layer.
```
Comment (mattdangerw) [inline keras_nlp/layers/fnet_encoder.py:147]: remove the `# LayerNorm layer.` and `# Feedforward layer.` comments, the function names communicate that fine.

**Review by mattdangerw — COMMENTED** [review]
Left some small comments. Once we resolve the default argument discussion this looks good to me.

File: keras_nlp/layers/fnet_encoder.py (hunk 9 above)
Comment (abheesht17) [reply keras_nlp/layers/fnet_encoder.py]: Right. I'll set the defaults to the ones used by `TransformerEncoder` 👍🏼

File: keras_nlp/layers/fnet_encoder.py (hunk 13 above)
Comment (abheesht17) [reply keras_nlp/layers/fnet_encoder.py]: Makes sense. Changed!

File: keras_nlp/layers/fnet_encoder.py (hunk 14 above)
Comment (abheesht17) [reply keras_nlp/layers/fnet_encoder.py]: 👍🏼

**Review by abheesht17 — COMMENTED** [review]
Thank you, @mattdangerw, for the review! I've changed all defaults to what they are in `TransformerEncoder` (including the initialisers).

**Review by mattdangerw — APPROVED** [review]
Thanks! This looks good to me.

**Comment by abheesht17** [comment]
Thank you, @mattdangerw! 🥳

File: keras_nlp/layers/fnet_encoder.py [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+    Note on padding: In the official FNet code, padding tokens are added to the
+    the input. However, the padding masks are deleted, i.e., mixing of
+    all tokens is done. This is because certain frequencies will be zeroed
+    out if we apply padding masks in every encoder layer. Hence, we don't
+    take padding mask as input in the call() function.
+
+    Args:
+        intermediate_dim: int, defaults to 3072. The hidden size of feedforward
+            network.
+        dropout: float, defaults to 0.1. The dropout value, applied in the
```
Comment (mattdangerw) [inline keras_nlp/layers/fnet_encoder.py:38]: oops just saw this, update the default in the docstrings too (or maybe just drop them), we generate docs we can include the default from the function signature

**Review by chenmoneygithub — APPROVED** [review]

File: keras_nlp/layers/fnet_encoder.py (hunk 15 above)
Comment (abheesht17) [reply keras_nlp/layers/fnet_encoder.py]: Ah, my bad. I've fixed it 👍🏼. Thanks for noticing!

**Review by fchollet — APPROVED** [review]
API LGTM.
